src/levenshtein.cr

```diff
@@ -155,4 +147,194 @@ module Levenshtein
   def self.find(name, all_names, tolerance = nil)
     Finder.find(name, all_names, tolerance)
   end
+
+  # Measures Levenshtein distance by filling Dynamic Programming Matrix
+  private def self.dynamic_matrix(string1 : String, string2 : String) : Int32
+    s_size = string1.size
+    l_size = string2.size
+
+    costs = Slice(Int32).new(s_size + 1) { |i| i }
+    last_cost = 0
+
+    if string1.single_byte_optimizable? && string2.single_byte_optimizable?
+      s = string1.to_unsafe
+      l = string2.to_unsafe
+
+      l_size.times do |i|
+        last_cost = i + 1
+
+        s_size.times do |j|
+          sub_cost = l[i] == s[j] ? 0 : 1
```

Comment:
`single_byte_optimizable?` means bytes higher than 0x7F should all compare equal, since they behave like the replacement character:

```crystal
Levenshtein.distance("\x81", "\x82").should eq(0)
Levenshtein.distance("\x82", "\x81").should eq(0)
Levenshtein.distance("\x81" * 33, String.new(Bytes.new(33) { |i| 0x80_u8 + i })).should eq(0)

# okay, as these use the `Char::Reader` overload instead
Levenshtein.distance("\x81", "\uFFFD").should eq(0)
Levenshtein.distance("\uFFFD", "\x81").should eq(0)
```

```suggestion
          l_i = {l[i], 0x80_u8}.min
          s_j = {s[j], 0x80_u8}.min
          sub_cost = l_i == s_j ? 0 : 1
```

This also means the `ascii_only?` branches could be relaxed to `single_byte_optimizable?`, operating over an alphabet of 129 "code points" where again bytes larger than 0x7F are mapped to 0x80.

Of course, most strings are already valid UTF-8 so this is a rather low priority issue. Feel free to ignore in this PR.

Reply:
You bring up an interesting point. I am not sure I ever really thought about what the Levenshtein distance of invalid strings would look like. But having said that the I don't even know what the point of this `single_byte_optimizable?` path, because if I am not mistaken no valid string will ever end up there, since if they are ascii it will call a different function and if it isn't then it wouldn't be single_byte_optimizable.

I think past me in an effort to optimize for speed looked into using unsafe pointers in non-ascii strings, to bring a speed boost to that specific scenario. And correct me if I am wrong, isn't this segment of code just providing a faster way to measure the Levenshtein distance of invalid strings? If that is the case then why even have it? 

Maybe getting rid of that 'if block' would be the cleanest solution.

Reply:
A string that consists only of ASCII characters plus invalid UTF-8 byte sequences, but not valid ones of 2 bytes or more (code point 0x80 or above), is `single_byte_optimizable?` but not `ascii_only?`.

Reply:
I guess my question is how often these occur in the wild. Is speeding up calculation of Levenshtein distance of strings that are `single_byte_optimizable?` but not `ascii_only?` even of value to anyone? Especially since the `Char::Reader` path is plenty fast to begin with, and handles the invalid bytes properly.